week6/week6_prepartion_database/Program.cs

```diff
@@ -0,0 +1,271 @@
+  var products = await connection.QueryAsync<Product>($"SELECT id, name, price,description FROM dapper.products where id={id}");
+  if (products.Count() == 0)
+  {
+    return Results.NotFound($" The  product with id:{id} does not exists in the database !!");
+  }
+  return Results.Ok(products);
+});
+
+/*Instead of a list, returns .FirstOrDefault*/
+app.MapGet("/firstordefault/{id}", async (IConfiguration configuration, int id) =>
+{
+  if (id < 0)
+  {
+    throw new Exception("id can not be negative");
+    // Results.BadRequest($" you entered id:{id},id can not be negative");
+  }
+  //#####
+  // how to checko if the input is letter instead of number or is not a number???
+  // var isNumeric = int.TryParse((string)id, out int value);
+  using var connection = new MySqlConnection(configuration.GetConnectionString("Default"));
+  var products = await connection.QueryAsync<Product>($"SELECT id, name, price,description FROM dapper.products where id={id}");
+  if (products.Count() == 0)
+  {
+    throw new Exception("product id not found");
+  }
+  return products.FirstOrDefault();
+});
+
+/*
+< summary >
+Endpoint for getgetting users:
+<questions>
+  1.Implement validation to return 400 if the name is null or empty string.
+  2.Bonus: Think what happens if you don’t send price in JSON?
+</questions>
+<example>
+postman input
+ <code>
+ {
+  "name": "pro name" ,
+  "price": 66.99 
+ }
+</code>
+</example>
+</summary>
+*/
+app.MapPost("/products", async (IConfiguration configuration, Product product) =>
+{
+  // it also works #####??
+  // if (product.Name == "" || product.Name == null || product.Price == null) // how to check null or empty for decimal???
+  // {
+  //   return Results.BadRequest($" name: '{product.Name}' or price: '{product.Price}' is not valid, please provide valid data !!");//400 badrequest
+  // }
+  if (string.IsNullOrEmpty(product.Name))
+  {
+    return Results.BadRequest($" name: '{product.Name}' or price: '{product.Price}' is not valid, please provide valid data !!");
+  }
+  await using var connection = new MySqlConnection(configuration.GetConnectionString("Default"));
+  var producPost = await connection.ExecuteAsync("INSERT INTO dapper.products (name, price, description) VALUES (@name, @price, @description)", product);// how to take the id that is recently posted??
+  return Results.Created($"/product/{producPost}", product); //  201 created
+});
+
+/*
+< summary >
+Endpoint for getgetting users:
+<questions>
+  1.Updating a product is done by handling a PUT endpoint on a /product/{id} endpoint.
+  2.What do you return from a PUT endpoint?
+</questions>
+<example>
+postman input
+ <code>
+ {
+  "name": "p name " ,
+  "price": 66.99 
+ }
+</code>
+</example>
+</summary>
+*/
+app.MapPut("/products/{id}", async (IConfiguration configuration, Product product, int id) =>
+{
+  if (product.Name == "" || product.Name == null || product.Name == " " || product.Price == null) // how to checko null or empty for decimal???
```

Comment:
`decimal` cannot be null ever. It will default to 0 if nothing is provided.
To make it nullable, you could write `decimal? Price {get;set;}`

Reply:
ok. thank you . got the point !

Reply:
`decimal Price {get;set;}` will have a default value of `0`
`decimal? Price {get;set;}` will have a default value of `null`

the `?` specifies that this property can be `null`